### PR TITLE
for pull request

### PR DESCRIPTION
**Pull request by TigerWolf**

### CODE REVIEW

File: data/input1.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+1, book, 12.49
```
Comment (TigerWolf) [inline data/input1.txt:1]: If these files are just for testing, you may want to put these into the rspec directory.

Comment (matcouto) [inline data/input1.txt:1]: hmmm, yeah. It makes more sense

File: spec/model/cart_spec.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    checkout_test(input_file(2))
+    cart.total.should eql 65.13 #Getting 65.13, instead of 65.15
+    cart.amount_sales_tax.should == 7.63 #Getting 7.63, instead of  7.65
+  end
+  it "calculates the total for the cart input 3" do
+    checkout_test(input_file(3))
+    cart.total.should eql 74.64 #Getting 74.64 instead of 74.68
+    cart.amount_sales_tax.should == 6.66 #Getting 6.66 instead of 6.70
+  end
+  def checkout_test(input)
```
Comment (matcouto) [inline spec/model/cart_spec.rb:62]: Is it ok to have helpers methods specific of a scenario like that? I mean, inside the describe block? Or I should have created an helper file and place those 2 method there?

File: service/cart_calculator.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+   def round_to(amount)
+    a = BigDecimal.new(amount.to_s)
+    a.mult(1, 2)
+    a.add(0, 4)
+    a.round(2)
+  end
+
+  def tax_to_be_applyed(product)
```
Comment (matcouto) [inline service/cart_calculator.rb:39]: Did you like the way I return the right tax to be applied? Using Decorator pattern. That's another thing I'd like to ask, I've seem many developers avoiding Design Patterns, why? I think patterns is independently of technology. I myself don't see any problem by bring some patterns that used to d on Java, to ruby.

File: service/cart_calculator.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    total
+  end
+  def apply_tax(tax_amount, amount_before_tax)
+      total = amount_before_tax + (amount_before_tax * tax_amount)
+      total = round_to(total)
+      @amount_sales_tax+= (total - amount_before_tax)
+      total
+  end
+
+   def round_to(amount)
```
Comment (matcouto) [inline service/cart_calculator.rb:32]: This little mate here is mocking me around. =) 
Most of the values of the examples in this challenge, it rounds correctly, but for a few, it misses a few cents =/
I even made a comment on the spec about this difference

File: service/cart_calculator.rb (hunk 3 above)
Comment (TigerWolf) [inline service/cart_calculator.rb:39]: I use Decorator pattern all the time. I use https://github.com/drapergem/draper in rails. I also really like to use service objects. One thing I would like to use more is Observer pattern, but its not really used a lot in rails as far as I have seen.

A really good article to read is: http://blog.codeclimate.com/blog/2012/10/17/7-ways-to-decompose-fat-activerecord-models/

One thing you will have to remember is ruby and rails prefers composition over inheritance. 

This is probably a good article on inheritance and composition in ruby, gives you some good guidelines too: http://ruby.learncodethehardway.org/book/ex44.html

Comment (TigerWolf) [inline service/cart_calculator.rb:39]: I also like your use of a NullObject http://robots.thoughtbot.com/rails-refactoring-example-introduce-null-object  with your BasicExempt.

File: service/cart_calculator.rb (hunk 4 above)
Comment (TigerWolf) [inline service/cart_calculator.rb:32]: The problem you are having is probably to do with integers and floats. You may need to convert the integer to a float using `to_f` before rounding it.

File: spec/model/cart_spec.rb (hunk 2 above)
Comment (TigerWolf) [inline spec/model/cart_spec.rb:62]: Check out shared examples: https://www.relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples

Comment (TigerWolf) [inline spec/model/cart_spec.rb:62]: Also if you are doing some sort of setup, use a `before` block:

https://www.relishapp.com/rspec/rspec-core/v/2-0/docs/hooks/before-and-after-hooks

```
before { 
   input.list_items.each do |item|
    cart.add_item_cart(item)
 end
 cart.checkout
 cart
}
```

Comment (TigerWolf) [inline spec/model/cart_spec.rb:62]: If you have several contexts you can just change the input file for each context: 

http://betterspecs.org/#context

```
before { 
   input.list_items.each do |item|
    cart.add_item_cart(item)
 end
 cart.checkout
 cart
}
context "cart 1" do
  let(:file) { Input.new("input1.txt") }

  it "something" do
    expect(cart.total_line_with_tax).to eq(54.63)
 end

end

context "cart 2" do
  let(:file) { Input.new("input2.txt") }

  it "something" do
    expect(cart.total_line_with_tax).to eq(54.63)
 end

end

```

Something like that ^. Fiddle around with it and ask any questions that come up.

Comment (TigerWolf) [inline spec/model/cart_spec.rb:62]: Also: http://betterspecs.org/ is awesome.

Comment (matcouto) [inline spec/model/cart_spec.rb:62]: Way better. Thanks for the tips. Will try to use it from now on ;)

**Pull request closed without merging**